modules/dnn/src/layers/affinegrid_layer.cpp

```diff
@@ -0,0 +1,325 @@
+        const bool is_float_ok = thetaType == CV_32F || thetaType == CV_64F || thetaType == CV_16BF || (allow_fp16 && thetaType == CV_16F) || thetaType == CV_16F;
+        CV_CheckType(thetaType, is_float_ok, "AffineGrid: theta must be a floating tensor");
+        CV_CheckType(inputs[1], inputs[1] == CV_64S || inputs[1] == CV_32S, "");
+
+        outputs.assign(std::max(1, requiredOutputs), CV_32F);
+        internals.assign(requiredInternals, CV_32F);
+    }
+
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
+    {
+        auto inKind = inputs_arr.kind();
+        auto outKind = outputs_arr.kind();
+        CV_Assert(inKind == _InputArray::STD_VECTOR_MAT || inKind == _InputArray::STD_VECTOR_UMAT);
+        CV_Assert(outKind == _InputArray::STD_VECTOR_MAT || outKind == _InputArray::STD_VECTOR_UMAT);
+
+        Mat theta, sz;
+        if (inKind == _InputArray::STD_VECTOR_MAT) {
+            theta = inputs_arr.getMat(0);
+            sz = inputs_arr.getMat(1);
+        } else {
+            UMat utheta = inputs_arr.getUMat(0);
+            UMat usz = inputs_arr.getUMat(1);
+            theta = utheta.getMat(ACCESS_READ);
+            sz = usz.getMat(ACCESS_READ);
+        }
+
+        CV_CheckTrue(theta.dims == 3, "theta must be 3D [N, r, c]");
+        const int N = theta.size[0];
+        const int r = theta.size[1];
+        const bool is3d = (r == 3);
+
+        int D = 1, H = 0, W = 0;
+        MatShape dummy;
+        resolveSizeAndShape(N, is3d, &sz, dummy, D, H, W);
+
+        CV_CheckGT(H, 0, "H must be > 0");
+        CV_CheckGT(W, 0, "W must be > 0");
+        if (is3d) CV_CheckGT(D, 0, "D must be > 0");
+
+        std::vector<int> outShape = is3d ? std::vector<int>{N, D, H, W, 3}
+                                         : std::vector<int>{N, H, W, 2};
+        MatShape fitShape(outShape.begin(), outShape.end());
+        Mat grid;
+        if (outKind == _InputArray::STD_VECTOR_MAT) {
+            std::vector<Mat>& outs = outputs_arr.getMatVecRef();
+            outs.resize(1);
+            outs[0].fit(fitShape, CV_32F);
+            grid = outs[0];
+        } else {
+            std::vector<UMat>& uouts = outputs_arr.getUMatVecRef();
+            uouts.resize(1);
+            uouts[0].fit(fitShape, CV_32F);
+            grid = uouts[0].getMat(ACCESS_WRITE);
+        }
+
+        auto computeLinspace = [&](int len, float& scale, float& delta){
+            if (align_corners) {
+                scale = len > 1 ? 2.f / float(len - 1) : 0.f;
+                delta = -1.f;
+            } else {
+                scale = 2.f / float(len);
+                delta = scale * 0.5f - 1.f;
+            }
+        };
+
+        float xs, xd, ys, yd, zs=0.f, zd=0.f;
+        computeLinspace(W, xs, xd);
+        computeLinspace(H, ys, yd);
+        if (is3d) computeLinspace(D, zs, zd);
+
+        auto dispatch2D = [&](int depth){
```

Comment:
why lambda functions are introduced here? You use each of them just once. Simply move the switch statements into `if (!is3d) { /* switch 2d */ } else { /* switch 3d */ }`